src/storage/txn/scheduler.rs

```diff
@@ -463,6 +440,106 @@ impl<E: Engine, L: LockManager> Scheduler<E, L> {
                 }
             })
             .unwrap();
+    }
+
+    /// Do some checking to decide whether fail fast or re-schedule.
+    fn fail_fast_or_reschedule(&self, cid: u64, cmd_ctx: Context) {
+        let sched = self.clone();
+        self.inner
+            .high_priority_pool
+            .pool
+            .spawn(async move {
+                let mut tctx = sched.inner.dequeue_task_context(cid);
+                // The task is not executed, so the task and cb should not be none.
+                let task = tctx.task.take().unwrap();
+                let cb = tctx.cb.take().unwrap();
+                let tag = task.cmd.tag();
+                match unsafe {
+                    with_tls_engine(|engine: &E| engine.precheck_write_with_ctx(&cmd_ctx))
+                } {
+                    Err(e) => {
+                        // Precheck failed, return err early.
+                        let pr = ProcessResult::Failed {
+                            err: StorageError::from(e),
+                        };
+                        Self::early_response(
+                            cid,
+                            cb,
+                            pr,
+                            tag,
+                            CommandStageKind::precheck_write_err,
+                        );
+                    }
+                    Ok(()) => {
+                        // Precheck passed, try to re-schedule the command with precheck disabled.
+                        SCHED_STAGE_COUNTER_VEC.get(tag).precheck_write_ok.inc();
+                        sched.schedule_command(task.cmd, cb, false);
+                    }
+                }
+            })
+            .unwrap();
+    }
+
+    /// Try to acquire all latches for the command and execute it.
+    /// Otherwise do some prechecking and then put it on the waiting queue.
+    fn schedule_command(&self, cmd: Command, callback: StorageCallback, fail_fast: bool) {
+        let cid = self.inner.gen_id();
+        let tracker = get_tls_tracker_token();
+        debug!("received new command"; "cid" => cid, "cmd" => ?cmd, "tracker" => ?tracker);
+        let tag = cmd.tag();
+        let priority_tag = get_priority_tag(cmd.priority());
+        SCHED_STAGE_COUNTER_VEC.get(tag).new.inc();
+        SCHED_COMMANDS_PRI_COUNTER_VEC_STATIC
+            .get(priority_tag)
+            .inc();
+
+        let cmd_ctx = cmd.ctx().clone();
+        let mut task_slot = self.inner.get_task_slot(cid);
+        let tctx = task_slot.entry(cid).or_insert_with(|| {
+            self.inner
+                .new_task_context(Task::new(cid, tracker, cmd), callback)
+        });
+
+        if fail_fast {
```

Comment:
I prefer to acquire latches like before, but when it fails to get all latches, spawn a background job to run the precheck (like the deadline one).

When the precheck fails, `try_own` the task, take the callback, and do the early return. In that way, I think the code will be simpler and the performance might also be better.